### PR TITLE
migrate buffer filter to v2 api

### PR DESCRIPTION
**Pull request by rshriram**
*title*: Convert buffer filter configuration into V2 API format
*Description*: Buffer filter in v2 format with tests
*Risk Level*: Low
*Testing*: unit tests

### CODE REVIEW

**Comment by rshriram** [comment]
cc @junr03

File: source/common/config/filter_json.cc [hunk 1]
```diff
@@ -274,5 +274,13 @@ void FilterJson::translateRouter(const Json::Object& json_router,
   router.set_start_child_span(json_router.getBoolean("start_child_span", false));
 }
 
+void FilterJson::translateBufferFilter(const Json::Object& json_buffer,
+                                       envoy::api::v2::filter::http::Buffer& buffer) {
+  json_buffer.validateSchema(Json::Schema::BUFFER_HTTP_FILTER_SCHEMA);
+
+  buffer.mutable_max_request_bytes()->set_value(json_buffer.getInteger("max_request_bytes"));
```
Comment (mattklein123) [inline source/common/config/filter_json.cc]: JSON_UTIL_SET_INTEGER ?

File: source/server/config/http/buffer.cc [hunk 2]
```diff
@@ -6,28 +6,44 @@
...
+                                       const std::string& stats_prefix, FactoryContext& context) {
+  ASSERT(buffer.has_max_request_bytes());
+  ASSERT(buffer.has_max_request_time());
 
   Http::BufferFilterConfigConstSharedPtr config(new Http::BufferFilterConfig{
       Http::BufferFilter::generateStats(stats_prefix, context.scope()),
-      static_cast<uint64_t>(json_config.getInteger("max_request_bytes")),
-      std::chrono::seconds(json_config.getInteger("max_request_time_s"))});
+      static_cast<uint64_t>(buffer.max_request_bytes().value()),
+      std::chrono::seconds(PROTOBUF_GET_MS_REQUIRED(buffer, max_request_time) / 1000)});
```
Comment (mattklein123) [inline source/server/config/http/buffer.cc]: nit: use duration_cast here (though if you can set seconds, can you also get seconds on the duration type?)

File: source/common/config/filter_json.h [hunk 3]
```diff
@@ -65,6 +66,15 @@ class FilterJson {
    */
   static void translateRouter(const Json::Object& json_router,
                               envoy::api::v2::filter::http::Router& router);
+
+  /**
+   * Translate a v1 JSON Buffer filter object to v2 envoy::api::v2::filter::http::Buffer.
+   * @param config source v1 JSON HTTP Buffer Filter object.
+   * @param buffer destination v2
```
Comment (htuch) [inline source/common/config/filter_json.h]: Name of parameters in docs and code are different.

File: source/server/config/http/buffer.cc (hunk 2 above)
Comment (rshriram) [reply source/server/config/http/buffer.cc]: The macro uses protobuf durationtoMilliseconds.. so we have to use some function or other. I can add a macro to get seconds or just do duration cast

Comment (rshriram) [reply source/server/config/http/buffer.cc]: created a new macro.

**Pull request merged**